File: Total-Hamming-Distance/THD.swift
class Solution {
    func totalHammingDistance(_ nums: [Int]) -> Int {
        var result = 0
        for i in 0..<nums.count {
            for ii in (i+1)..<nums.count{
                result += diffHam(nums[i],nums[ii])
            }
        }
        return result
    }

    func diffHam(_ n1:Int, _ n2:Int) -> Int {
        var a = n1
        var b = n2
        var result = 0
        while(a != 0 || b != 0){
            if a%2 != b%2{
                result += 1
            }
            (a,b) = (a/2,b/2)
        }
        return result
    }

}

File: Total-Hamming-Distance/THD.hs
diffHam :: (Integral a) => a -> a -> a
diffHam a b
  | a == 0 && b == 0 = 0
  | otherwise = let rr = if (mod a 2) /= (mod b 2)
                      then 1
                      else 0
                in rr + diffHam (div a 2) (div b 2)

totalHammingDistance :: (Integral a) => [a] -> a
totalHammingDistance []     = 0
totalHammingDistance (x:xs) = (sum $ map (diffHam x) xs) + totalHammingDistance xs
